app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/liveview/LiveviewIoThread.java

```diff
@@ -0,0 +1,220 @@
+            ParcelUuid uuids[] = btDevice.getUuids();
+            if (uuids == null) {
+                return false;
+            }
+            for (ParcelUuid uuid : uuids) {
+                LOG.info("found service UUID " + uuid);
+            }
+            mBtSocket = btDevice.createRfcommSocketToServiceRecord(uuids[0].getUuid());
+            mBtSocket.connect();
+            mInStream = mBtSocket.getInputStream();
+            mOutStream = mBtSocket.getOutputStream();
+            setUpdateState(GBDevice.State.CONNECTED);
+        } catch (IOException e) {
+            e.printStackTrace();
+            LOG.error("Server socket cannot be started");
+            setUpdateState(originalState);
+            mInStream = null;
+            mOutStream = null;
+            mBtSocket = null;
+            return false;
+        }
+
+        write(mLiveviewProtocol.encodeSetTime());
+        setUpdateState(GBDevice.State.INITIALIZED);
+
+        return true;
+    }
+
+    private void setUpdateState(GBDevice.State state) {
+        gbDevice.setState(state);
+        gbDevice.sendDeviceUpdateIntent(getContext());
+    }
+
+    private byte[] parseIncoming() throws IOException {
+        ByteArrayOutputStream msgStream = new ByteArrayOutputStream();
+
+        int needRead = 1;
+        ReaderState state = ReaderState.ID;
+        do {
+
+            byte read = -1;
+            read = (byte) mInStream.read();
+
+            if (read == -1) {
+                LOG.error("Invalid message received (length=" + msgStream.size() + ")");
+            }
+            needRead--;
+            msgStream.write(read);
+            if (needRead == 0) {
+                switch (state) {
+                    case ID:
+                        state = ReaderState.HEADER_LEN;
+                        needRead = 1;
+                        break;
+                    case HEADER_LEN:
+                        state = ReaderState.HEADER;
+                        needRead = read;
+                        break;
+                    case HEADER:
+                        int payloadSize = getLastInt(msgStream);
+                        state = ReaderState.PAYLOAD;
+                        needRead = payloadSize;
+                        break;
+                    default:
+                        break;
+                }
+            }
+        } while (needRead > 0);
+        byte[] msgArray = msgStream.toByteArray();
+        LOG.debug("received: " + GB.hexdump(msgArray, 0, msgArray.length));
+        return msgArray;
+    }
+
+
+    /**
+     * Enumeration containing the possible internal status of the reader.
+     */
+    private enum ReaderState {
+        ID, HEADER_LEN, HEADER, PAYLOAD;
+    }
+
+    private int getLastInt(ByteArrayOutputStream stream) {
+        byte[] array = stream.toByteArray();
+        ByteBuffer buffer = ByteBuffer.wrap(array, array.length - 4, 4);
+        buffer.order(ByteOrder.BIG_ENDIAN);
```

Comment:
I'd add a constant for the byte order in LiveviewConstants and use that everywhere, but I agree it's rather unlikely that this will ever change ;-)